backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -0,0 +1,188 @@
+        cell_counts_df = pd.concat([cell_counts_df, to_attach], axis=0)
+        cell_counts_df_rollup = rollup_across_cell_type_descendants(cell_counts_df)
+        cell_counts_df = cell_counts_df.set_index("cell_type_ontology_term_id")["n_cells"]
+        cell_counts_df_rollup = cell_counts_df_rollup.set_index("cell_type_ontology_term_id")["n_cells"]
+        return cell_counts_df, cell_counts_df_rollup
+
+    def _delete_unknown_terms_from_ontology_graph(self, node):
+        """
+        Deletes nodes that have no name in the CL ontology from the ontology graph.
+
+        Arguments
+        ---------
+        node: dict
+            A node of the ontology graph. At the top-level, this should be the root node of the ontology graph.
+        """
+        new_children = []
+        for child in node.get("children", []):
+            unknown = child["name"].startswith("CL:")
+            if not unknown:
+                new_children.append(child)
+        if len(new_children) > 0:
+            node["children"] = new_children
+        elif "children" in node:
+            del node["children"]
+
+        for child in node.get("children", []):
+            self._delete_unknown_terms_from_ontology_graph(child)
+
+    def _initialize_id_to_name(self):
+        """
+        Initializes a dictionary that maps cell type ontology term ids to their names.
+
+        Returns
+        -------
+        id_to_name: dict
+            A dictionary that maps cell type ontology term ids to their names.
+        all_cell_type_ids: list
+            A list of all cell type ontology term ids.
+        """
+        id_to_name = {}
+        for term in self.ontology.terms():
+            if term.id.startswith("CL:"):
+                id_to_name[term.id] = term.name
+        all_cell_type_ids = list(id_to_name.keys())
+        return id_to_name, all_cell_type_ids
+
+    def _traverse_ontology_with_counting(self, node, traverse_node_counter=None, all_unique_nodes=None):
+        """
+        This function traverses the cell ontology and builds a nested dictionary representation of the tree.
+        It also counts the number of times a node has been visited and adds a suffix to the node id to make it unique.
+        The suffix is the number of times the node had been visited prior.
+
+        Arguments
+        ---------
+        node: pronto.Term
+            A node of the cell ontology. This should be the root node of the ontology at the top-level.
+        traverse_node_counter: dict, optional, default=None
+            A dictionary that maps cell type ontology term ids to the number of times they have been visited.
+        all_unique_nodes: set, optional, default=None
+            A set of all unique cell type ontology term ids that have been visited.
+
+        Returns
+        -------
+        ontology_graph: dict
+            A nested dictionary representation of the cell ontology tree.
+        """
+        if traverse_node_counter is None:
```

Comment:
this is kind of a nit, but i think it would be a bit better to remove this block here, instead initialize the call in `OntologyTreeBuilder` to set `traverse_node_counter = {}` and `all_unique_nodes = set()`, and require the parameters to be set rather than having a default of `None`

Reply:
That's what it was, but I think it's an anti-pattern. `traverse_node_counter` is an argument only because it's required to be passed to recursive calls of the traversal function. As long as the function documents what it returns, then there's no need to initialize it outside the function call.